### PR TITLE
build: modest Android improvements

### PR DESCRIPTION
**Pull request by icota**
I've been trying to build for Android on different OSes/Gitian with varying success. Build system is quite the beast and sometimes it doesn't get it right. To make sure it does these three little tweaks make the Android build more robust:

- disable D-Bus (Android doesn't support it and has its own way to trigger notifications)
- don't flag `-lpthread` when linking Boost, [Bionic has built-in support](https://stackoverflow.com/questions/30801752/android-ndk-and-pthread)
- ~~add `-static-libstdc++` to linker flags. This avoids having to bundle `libc++_shared` with CLI apps, still necessary with `bitcoin-qt` though (thanks Sjors)~~

I think these are small and fairly straightforward so I put them all into this one PR.

### CODE REVIEW

File: build-aux/m4/bitcoin_qt.m4 [hunk 1]
```diff
@@ -76,6 +76,13 @@ AC_DEFUN([BITCOIN_QT_INIT],[
     [use_dbus=$withval],
     [use_dbus=auto])
 
+  dnl Android doesn't support D-Bus and certainly doesn't use it for notifications
```
Comment (laanwj) [inline build-aux/m4/bitcoin_qt.m4]: Please update the help message above for `--with-qtdbus`'s default

Comment (icota) [reply build-aux/m4/bitcoin_qt.m4]: I was considering that but I feel that the configuration should be as independent as possible from the depends build or whatever libs we decide to use. For example a user might point to prebuilt Qt for Android libs (with QtDbus) and if that breaks the build he is going to blame the `configure` script.

There's definitely no use for D-Bus on Android so I think it okay to disable even if the lib is found.

File: build-aux/m4/ax_boost_thread.m4 [hunk 2]
```diff
@@ -71,6 +71,8 @@ AC_DEFUN([AX_BOOST_THREAD],
                  CXXFLAGS="-pthreads $CXXFLAGS"
              elif test "x$host_os" = "xmingw32" ; then
                  CXXFLAGS="-mthreads $CXXFLAGS"
+             elif test "x$host_os" = "xlinux-android" ; then
+                 CXXFLAGS="$CXXFLAGS"
```
Comment (luke-jr) [inline build-aux/m4/ax_boost_thread.m4]: Does `-pthread` actually *fail* targeting Android?

File: build-aux/m4/bitcoin_qt.m4 (hunk 1 above)
Comment (luke-jr) [reply build-aux/m4/bitcoin_qt.m4]: Maybe some Android fork might add D-Bus, though? At least allow --with-dbus to override...

File: configure.ac [hunk 3]
```diff
@@ -599,6 +599,7 @@ case $host in
      dnl make sure android stays above linux for hosts like *linux-android*
      TARGET_OS=android
      LEVELDB_TARGET_FLAGS="-DOS_ANDROID"
+     LDFLAGS="$LDFLAGS -static-libstdc++"
```
Comment (luke-jr) [inline configure.ac]: Better to bundle a shared library than to link statically.

**Review by luke-jr — CHANGES_REQUESTED** [review]

File: build-aux/m4/ax_boost_thread.m4 (hunk 2 above)
Comment (icota) [reply build-aux/m4/ax_boost_thread.m4]: Yes it does, see [here](https://github.com/protocolbuffers/protobuf/issues/1373). I'm not sure why it worked before but right now building the master branch on Linux is broken.

File: configure.ac (hunk 3 above)
Comment (icota) [reply configure.ac]: Google [recommends static](https://developer.android.com/ndk/guides/cpp-support#important_considerations).

Comment (luke-jr) [reply configure.ac]: "If all of your application's native code is contained in a single shared library, ..."

Which it isn't.

Comment (icota) [reply configure.ac]: I assumed that most people will want `bitcoind` without the utils. I don't have strong opinions about this so if you think bundling more than one bin + shared lib is a more common usecase I'll drop this commit.

Comment (luke-jr) [reply configure.ac]: More likely `bitcoin-qt`!

**Comment by greenaddress** [comment]
Seems to improve some issues found on android but not all - on top of doing checks for "xlinux-android" it should also do checks for "xlinux-androideabi" otherwise you cut off armv7

I'm doing some testing on this, will report my findings

**Comment by greenaddress** [comment]
@icota I can confirm that with xlinux-androideabi as well as xlinux-android all android ndk builds work fine for me/abcore - see https://github.com/greenaddress/bitcoin_ndk/blob/master/0001-android-patches-1bc9988993ee84bc814e5a7f33cc90f670a19f6a.patch

**Comment by icota** [comment]
Thanks @greenaddress! In that case for the case statement we should probably use wildcard(s) to look for `xlinux-android*` or even `*android*`. 

I'm not sure if the `test "x$host_os" = "x...` bits allow for wildcards, if not I'd convert them to use `case` since it's neater. I'll investigate.

**Comment by icota** [comment]
Rebased and in `ax_boost_thread.m4` converted `if test` to `case` statements for easier matching of Android flavours.

File: build-aux/m4/ax_boost_thread.m4 [hunk 4]
```diff
@@ -67,13 +67,24 @@ AC_DEFUN([AX_BOOST_THREAD],
...
              CXXFLAGS_SAVE=$CXXFLAGS
 
-             if test "x$host_os" = "xsolaris" ; then
-                 CXXFLAGS="-pthreads $CXXFLAGS"
-             elif test "x$host_os" = "xmingw32" ; then
-                 CXXFLAGS="-mthreads $CXXFLAGS"
-             else
-                CXXFLAGS="-pthread $CXXFLAGS"
-             fi
+             case "x$host_os" in
```
Comment (fanquake) [inline build-aux/m4/ax_boost_thread.m4]: Have you sent these changes upstream? If so, can you link to the upstream commits/PR. I'd rather not modify the macros in our tree unless absolutely necessary.

Comment (icota) [reply build-aux/m4/ax_boost_thread.m4]: To be honest I didn't even know there was an upstream. I'll send a patch file (or whatever GNU people do nowadays) and report back here.

Comment (fanquake) [reply build-aux/m4/ax_boost_thread.m4]: Great, thanks. You should be able to open a PR against the [autoconf-archive](https://github.com/autoconf-archive/autoconf-archive) repo on GitHub.

Comment (icota) [reply build-aux/m4/ax_boost_thread.m4]: [Done](https://github.com/autoconf-archive/autoconf-archive/pull/211).

Comment (icota) [reply build-aux/m4/ax_boost_thread.m4]: [Merged upstream](https://github.com/autoconf-archive/autoconf-archive/commit/3eaf1697e4734e2ed2e9960fa755733e01042223).

**Comment by MarcoFalke** [comment]
Concept ACK

**Comment by fanquake** [comment]
@icota If you'd like to move forward here, can you do the following. 77a347c7115125263561ab76389b5a47e1fb5a4a and dfdc8b3df58024884049db057a3d9d1312aa2722 should be combined into a single commit, that is actually just pulling in the latest version of the macro from upstream (looks like all the changes you made have been upstreamed), with an appropriate commit message. See 9c5bef450f7bf4dd4307cae8bce76dc131643574 in #19558 for an example of what I mean. I also think you can squash the other two commits.

**Comment by icota** [comment]
Thanks @fanquake. I've rebased and done exactly that.

Did a test Android ARM64 full build which completed successfully. I ran a normal native build as well and it also completed without error.

**Comment by fanquake** [comment]
> Thanks @fanquake. I've rebased and done exactly that.

@icota Thanks for following up, this looks good now. Just waiting for the gitian build.

**Review by fanquake — APPROVED** [review]
ACK 366913e307d2dc13bc00d6bf7b6b2426c359ac30

**Pull request merged**